apps/performance-tests/src/util/TestUtilities.ts

```diff
@@ -3,10 +3,40 @@
  * See LICENSE.md in the project root for license terms and full copyright notice.
  *--------------------------------------------------------------------------------------------*/
 
+interface AdvancedRunProps<TContext> {
+  /** Callback to run before the test that should produce the context required for the test. */
+  setup(): TContext | Promise<TContext>;
+
+  /** Test function to run and measure. */
+  test(x: TContext): void | Promise<void>;
+
+  /** Callback that cleans up the context produced by the "before" callback. */
+  cleanup?: (x: TContext) => void | Promise<void>;
+}
+
 /** Runs a test and passes information about it to the TestReporter. */
-export function run(desc: string, callback: (() => void) | (() => Promise<void>)) {
+export function run(desc: string, callback: () => void | Promise<void>): void;
+export function run<T>(desc: string, props: AdvancedRunProps<T>): void;
+export function run<T>(desc: string, callbackOrProps: (() => void | Promise<void>) | AdvancedRunProps<T>): void {
   it(desc, async function () {
-    this.test!.ctx!.testReporterOnTestStart();
-    await callback();
+    if (typeof callbackOrProps === "function") {
+      this.test!.ctx!.reporter.onTestStart();
+      await callbackOrProps();
+      return;
+    }
+
+    let value: T;
+    try {
+      value = await callbackOrProps.setup();
+    } finally {
+      this.test!.ctx!.reporter.onTestStart();
+    }
```

Comment:
I don't see a reason to place this into `try/finally` - if the setup fails we're going to throw, so why even call `onTestStart`?

Reply:
If `setup()` throws an error we will only be able to see the test failure at the end:
```
 models tree
    ✔ initial (Baytown) (58.55 ms)
    ✔ full (Baytown) (1503.11 ms)

Status | Test                          | Duration   | Blocks                                           
-------------------------------------------------------------------------------------------------------
PASS   | models tree initial (Baytown) | 58.55 ms   | count: 0                                         
PASS   | models tree full (Baytown)    | 1503.11 ms | count: 3, max: 130.00, p95: 130.00, median: 71.00

flat 50k elements list:
Error
    at Object.setup (src/stateless.test.ts:36:11)
    at Context.<anonymous> (src/util/TestUtilities.ts:28:41)
    at process.processImmediate (node:internal/timers:476:21)
```

Reply:
makes sense, thanks